tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_sqs.py

```diff
@@ -1042,6 +1043,182 @@ def test_sqs_event_source_mapping(
+            )
+            assert sum([len(batch) for batch in batches]) == 15
+            return [message for batch in batches for message in batch]
+
+        events = retry(get_msg_from_q, retries=15, sleep=5)
+        snapshot.match("Records", events)
+
+    # FIXME: this fails due to ESM not correctly collecting and sending batches
+    # where size exceeds 10 messages.
+    @markers.snapshot.skip_snapshot_verify(paths=["$..total_batches_received"])
+    @markers.aws.validated
+    def test_sqs_event_source_mapping_batching_reserved_concurrency(
+        self,
+        create_lambda_function,
+        sqs_create_queue,
+        sqs_get_queue_arn,
+        lambda_su_role,
+        snapshot,
+        cleanups,
+        aws_client,
+    ):
+        snapshot.add_transformer(snapshot.transform.sqs_api())
+        snapshot.add_transformer(SortingTransformer("Records", lambda s: s["body"]), priority=-1)
+
+        destination_queue_name = f"destination-queue-{short_uid()}"
+        function_name = f"lambda_func-{short_uid()}"
+        source_queue_name = f"source-queue-{short_uid()}"
+        mapping_uuid = None
+
+        destination_queue_url = sqs_create_queue(QueueName=destination_queue_name)
+        create_lambda_function(
+            func_name=function_name,
+            handler_file=TEST_LAMBDA_EVENT_SOURCE_MAPPING_SEND_MESSAGE,
+            runtime=Runtime.python3_12,
+            envvars={"SQS_QUEUE_URL": destination_queue_url},
+            role=lambda_su_role,
+        )
+
+        # Prevent more than 2 Lambdas from being spun up at a time
+        put_concurrency_resp = aws_client.lambda_.put_function_concurrency(
+            FunctionName=function_name, ReservedConcurrentExecutions=2
+        )
+        snapshot.match("put_concurrency_resp", put_concurrency_resp)
+
+        queue_url = sqs_create_queue(QueueName=source_queue_name)
+        queue_arn = sqs_get_queue_arn(queue_url)
+
+        create_event_source_mapping_response = aws_client.lambda_.create_event_source_mapping(
+            EventSourceArn=queue_arn,
+            FunctionName=function_name,
+            MaximumBatchingWindowInSeconds=10,
+            BatchSize=20,
+            ScalingConfig={
+                "MaximumConcurrency": 2
+            },  # Prevent more than 2 concurrent SQS workers from being spun up at a time
+        )
+        mapping_uuid = create_event_source_mapping_response["UUID"]
+        cleanups.append(lambda: aws_client.lambda_.delete_event_source_mapping(UUID=mapping_uuid))
+        snapshot.match("create-event-source-mapping-response", create_event_source_mapping_response)
+        _await_event_source_mapping_enabled(aws_client.lambda_, mapping_uuid)
+
+        for b in range(3):
+            aws_client.sqs.send_message_batch(
+                QueueUrl=queue_url,
+                Entries=[{"Id": f"{i}-{b}", "MessageBody": f"{i}-{b}-message"} for i in range(10)],
+            )
+
+        batches = []
+
+        def get_msg_from_q():
```

Comment:
idea for future: It feels we re-implement this pattern all over the place in different ways and should consider using an appropriate helper/fixture 🙈